crates/curp/src/server/cmd_worker/mod.rs

```diff
@@ -1,269 +1,244 @@
-                .remove(PoolEntry::new(entry.propose_id, Arc::clone(cmd)));
-            if curp.is_leader() {
-                ucp.lock()
-                    .remove(PoolEntry::new(entry.propose_id, Arc::clone(cmd)));
+            AfterSyncCmd::new(
+                cmd.as_ref(),
+                // If the response sender is absent, it indicates that a new leader
+                // has been elected, and the entry has been recovered from the log
+                // or the speculative pool. In such cases, these entries needs to
+                // be re-executed.
+                tx.as_ref().map_or(true, |t| t.is_conflict()),
+            )
+        })
+        .collect();
+
+    match ce.after_sync(cmds, highest_index).await {
+        Ok(resps) => {
+            for ((asr, er_opt), tx) in resps
+                .into_iter()
+                .zip(resp_txs)
+                .filter_map(|(resp, tx_opt)| tx_opt.as_ref().map(|tx| (resp, tx)))
+            {
+                if let Some(er) = er_opt {
+                    tx.send_propose(ProposeResponse::new_result::<C>(&Ok(er), true));
+                }
+                tx.send_synced(SyncedResponse::new_result::<C>(&Ok(asr)));
             }
-            debug!("{id} cmd({}) after sync is called", entry.propose_id);
-            asr_ok
         }
-        EntryData::Shutdown => {
-            curp.task_manager().cluster_shutdown();
-            if curp.is_leader() {
-                curp.task_manager().mark_leader_notified();
-            }
-            if let Err(e) = ce.set_last_applied(entry.index) {
-                error!("failed to set last_applied, {e}");
+        Err(e) => {
+            for tx in resp_txs.flatten() {
+                tx.send_synced(SyncedResponse::new_result::<C>(&Err(e.clone())));
             }
-            cb.write().notify_shutdown();
-            true
         }
-        EntryData::ConfChange(ref conf_change) => {
-            if let Err(e) = ce.set_last_applied(entry.index) {
-                error!("failed to set last_applied, {e}");
-                return false;
-            }
-            let change = conf_change.first().unwrap_or_else(|| {
-                unreachable!("conf change should always have at least one change")
-            });
-            let shutdown_self =
-                change.change_type() == ConfChangeType::Remove && change.node_id == id;
-            cb.write().insert_conf(entry.propose_id);
-            sp.lock()
-                .remove(PoolEntry::new(entry.propose_id, conf_change.clone()));
-            if curp.is_leader() {
-                ucp.lock()
-                    .remove(PoolEntry::new(entry.propose_id, conf_change.clone()));
+    }
+
+    for (entry, _) in &cmd_entries {
+        curp.trigger(&entry.propose_id);
+        ce.trigger(entry.inflight_id());
+    }
+    let mut sp_l = sp.lock();
+    let mut ucp_l = ucp.lock();
+    for (entry, _) in cmd_entries {
+        remove_from_sp_ucp(&mut sp_l, &mut ucp_l, &entry);
+    }
```

Comment:
I suggest to moving all the lock operations into the `remove_from_sp_ucp` function.

Reply:
done